runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -869,20 +694,98 @@ static iree_status_t iree_hal_hip_device_create_channel(
     iree_hal_command_buffer_t** out_command_buffer) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
   iree_hal_hip_device_t* device = iree_hal_hip_device_cast(base_device);
-  IREE_RETURN_IF_ERROR(
-      iree_hal_hip_set_context(device->hip_symbols, device->hip_context));
-
-  return iree_hal_hip_stream_command_buffer_create(
-      iree_hal_device_allocator(base_device), device->hip_symbols,
-      device->nccl_symbols, device->hip_context, device->tracing_context, mode,
-      command_categories, binding_capacity, device->hip_dispatch_stream,
-      &device->block_pool, device->host_allocator, out_command_buffer);
+
+  iree_hal_command_buffer_t* buffers[IREE_HAL_MAX_QUEUES];
+  memset(buffers, 0x00,
+         sizeof(iree_hal_command_buffer_t*) * IREE_HAL_MAX_QUEUES);
+  if (queue_affinity == 0) {
+    queue_affinity = IREE_HAL_QUEUE_AFFINITY_ANY;
+  }
+  queue_affinity =
+      queue_affinity & ~(IREE_HAL_QUEUE_AFFINITY_ANY << device->topology.count);
+
+  iree_status_t status = iree_ok_status();
+  iree_host_size_t idx = 0;
+  iree_host_size_t cb_num = 0;
+  iree_hal_queue_affinity_t affinity = queue_affinity;
+  while (iree_status_is_ok(status) && affinity) {
+    iree_host_size_t next_idx_offset =
+        iree_math_count_trailing_zeros_u64(affinity);
+    idx += next_idx_offset;
+    affinity >>= next_idx_offset + 1;
+    status = IREE_HIP_RESULT_TO_STATUS(
+        device->hip_symbols,
+        hipCtxPushCurrent(device->topology.devices[idx].hip_context));
+    if (!iree_status_is_ok(status)) {
+      break;
+    }
+    switch (type) {
+      case IREE_HAL_HIP_DEVICE_COMMAND_BUFFER_TYPE_STREAM:
+        status = iree_status_join(
+            status,
+            iree_hal_hip_stream_command_buffer_create(
+                iree_hal_device_allocator(base_device), device->hip_symbols,
+                device->nccl_symbols,
+                device->topology.devices[idx].tracing_context, mode,
+                command_categories, (iree_hal_queue_affinity_t)1 << idx,
+                binding_capacity,
+                device->topology.devices[idx].hip_dispatch_stream,
+                &device->block_pool, device->host_allocator, &buffers[cb_num]));
+        break;
+      case IREE_HAL_HIP_DEVICE_COMMAND_BUFFER_TYPE_GRAPH:
+        status = iree_status_join(
+            status,
+            iree_hal_hip_graph_command_buffer_create(
+                iree_hal_device_allocator(base_device), device->hip_symbols,
+                device->topology.devices[idx].tracing_context,
+                device->topology.devices[idx].hip_context, mode,
+                command_categories, (iree_hal_queue_affinity_t)1 << idx,
+                binding_capacity, &device->block_pool, device->host_allocator,
+                &buffers[cb_num]));
+        break;
+    }
+    status =
+        IREE_HIP_RESULT_TO_STATUS(device->hip_symbols, hipCtxPopCurrent(NULL));
+    ++idx;
+    ++cb_num;
+  }
+
+  if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
```

Comment:
this is non-standard - do an `if(ok)` and do your command_buffer_create and then clean up afterwards

this right now will leak the buffers if the command_buffer_create fails and is another early exit from the function that makes it harder to follow

```c
if (status is ok) {
  status = iree_hal_hip_multi_queue_command_buffer_create(...)
}
if (!status is ok) {
  release all buffers
}
end zone;
return;
```

related, you also want to make iree_hal_hip_multi_queue_command_buffer_create retain the buffers passed in - the only time the caller should be retaining for the callee is if the entire list ownership is transferred - here it's not and that makes it confusing (buffers is stack allocated here and then copied out in iree_hal_hip_multi_queue_command_buffer_create, but the ref ownership is transferred of the individual elements). a few extra ref counts is worth the consistency - you have to loop over the items anyway and they'll be in cache from the retains out here anyway.

then you do
```
if (status is ok) {
  status = create
}
unconditionally release all buffers regardless of status;
end zone;
return;